system/diagnostic_graph_aggregator/src/node/recovery.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 The Autoware Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "recovery.hpp"
+
+#include <algorithm>
+
+namespace diagnostic_graph_aggregator
+{
+
+std::string level_to_string(DiagnosticLevel level)
+{
+  switch (level) {
+    case DiagnosticStatus::OK:
+      return "OK";
+    case DiagnosticStatus::WARN:
+      return "WARN";
+    case DiagnosticStatus::ERROR:
+      return "ERROR";
+    case DiagnosticStatus::STALE:
+      return "STALE";
+  }
+  return "UNKNOWN";
+}
+
+
+RecoveryNode::RecoveryNode() : Node("recovery")
+{
+  using std::placeholders::_1;
+  const auto qos_graph = rclcpp::QoS(1);
+  const auto qos_aw_state = rclcpp::QoS(1);
+  const auto qos_mrm_state = rclcpp::QoS(1);
+
+  const auto callback_graph = std::bind(&RecoveryNode::on_graph, this, _1);
+  sub_graph_ = create_subscription<DiagnosticGraph>("/diagnostics_graph", qos_graph, callback_graph);
+  const auto callback_aw_state = std::bind(&RecoveryNode::on_aw_state, this, _1);
+  sub_aw_state_ = create_subscription<AutowareState>("/autoware/state", qos_aw_state, callback_aw_state);
+  const auto callback_mrm_state = std::bind(&RecoveryNode::on_mrm_state, this, _1);
+  sub_mrm_state_ = create_subscription<MrmState>("/system/fail_safe/mrm_state",
+    qos_mrm_state, callback_mrm_state);
+  callback_group_ = create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  srv_clear_mrm_ =  create_client<std_srvs::srv::Trigger>("/system/clear_mrm",
+    rmw_qos_profile_services_default, callback_group_);
+
+  fatal_error_ = false;
+  mrm_occur_ = false;
+}
+
+void RecoveryNode::on_graph(const DiagnosticGraph::ConstSharedPtr msg)
+{
+  bool autonomous_available = false;
+  for (const auto & node : msg->nodes) {
+    if (node.status.name == "/autoware/modes/autonomous") {
+      autonomous_available = node.status.level == DiagnosticStatus::OK;
+    }
+    // aggregate non-recoverable error
+    if (node.status.name == "/autoware/fatal_error/autonomous_available") {
+      if (node.status.level != DiagnosticStatus::OK){
+        fatal_error_ = true;
+      }
+    }
+  }
+
+  // 1. Not emergency
+  // 2. Non-recovoerable errors have not happened
+  // 3. on MRM
+  if (autonomous_available && !fatal_error_ && mrm_occur_){
+    clear_mrm();
+  }
+}
+
+void RecoveryNode::on_aw_state(const AutowareState::ConstSharedPtr msg){
+  if (msg->state != AutowareState::DRIVING)
+    fatal_error_ = false;
```

Comment:
@saka1-s 
I think the autoware state will be ARRIVED_GOAL if the vehicle stops with pull over MRM.
If pull over succeeds after a fatal error, it will recover, so if you want to ignore the startup diag, you should reset fatal_error_ before the first automatic operation.

Reply:
@mkuri 
Thank you. I have confirmed that your point is correct.
As a fix, I would like to introduce the mrm_by_fatal_error_ flag.
This variable monitors whether fatal_error_ has occurred and the mrm state has actually changed.
If the mrm state transitions from a state other than normal to normal, the mrm_by_fatal_error_ flag is reset.
Because I found another problem.
When the user cancels MRM after a fatal_error has occurred. If fatal_error_ is not reset in this case, it will not be automatically cancelled the next time MRM occurs.
I have confirmed that this method will also solve the problem you pointed out.